base/compiler/ssair/EscapeAnalysis/EscapeAnalysis.jl

```diff
@@ -0,0 +1,1604 @@
+    (; xidx, yidx) = change
+    xroot = find_root!(estate.aliasset, xidx)
+    yroot = find_root!(estate.aliasset, yidx)
+    if xroot ≠ yroot
+        union!(estate.aliasset, xroot, yroot)
+        xinfo = estate.escapes[xidx]
+        yinfo = estate.escapes[yidx]
+        xyinfo = xinfo ⊔ yinfo
+        estate.escapes[xidx] = xyinfo
+        estate.escapes[yidx] = xyinfo
+        return true
+    end
+    return false
+end
+
+function add_escape_change!(astate::AnalysisState, @nospecialize(x), xinfo::EscapeLattice)
+    xinfo === ⊥ && return nothing # performance optimization
+    xidx = iridx(x, astate.estate)
+    if xidx !== nothing
+        if !isbitstype(widenconst(argextype(x, astate.ir)))
+            push!(astate.changes, EscapeChange(xidx, xinfo))
+        end
+    end
+    return nothing
+end
+
+function add_liveness_change!(astate::AnalysisState, @nospecialize(x), livepc::Int)
+    xidx = iridx(x, astate.estate)
+    if xidx !== nothing
+        if !isbitstype(widenconst(argextype(x, astate.ir)))
+            push!(astate.changes, LivenessChange(xidx, livepc))
+        end
+    end
+    return nothing
+end
+
+function add_alias_change!(astate::AnalysisState, @nospecialize(x), @nospecialize(y))
+    if isa(x, GlobalRef)
+        return add_escape_change!(astate, y, ⊤)
+    elseif isa(y, GlobalRef)
+        return add_escape_change!(astate, x, ⊤)
+    end
+    estate = astate.estate
+    xidx = iridx(x, estate)
+    yidx = iridx(y, estate)
+    if xidx !== nothing && yidx !== nothing && !isaliased(xidx, yidx, astate.estate)
+        pushfirst!(astate.changes, AliasChange(xidx, yidx))
+    end
+    return nothing
+end
+
+function escape_edges!(astate::AnalysisState, pc::Int, edges::Vector{Any})
+    ret = SSAValue(pc)
+    for i in 1:length(edges)
+        if isassigned(edges, i)
+            v = edges[i]
+            add_alias_change!(astate, ret, v)
+        end
+    end
+end
+
+function escape_val_ifdefined!(astate::AnalysisState, pc::Int, x)
+    if isdefined(x, :val)
+        escape_val!(astate, pc, x.val)
+    end
+end
+
+function escape_val!(astate::AnalysisState, pc::Int, @nospecialize(val))
+    ret = SSAValue(pc)
+    add_alias_change!(astate, ret, val)
+end
+
+# NOTE if we don't maintain the alias set that is separated from the lattice state, we can do
+# soemthing like below: it essentially incorporates forward escape propagation in our default
+# backward propagation, and leads to inefficient convergence that requires more iterations
+# # lhs = rhs: propagate escape information of `rhs` to `lhs`
+# function escape_alias!(astate::AnalysisState, @nospecialize(lhs), @nospecialize(rhs))
+#     if isa(rhs, SSAValue) || isa(rhs, Argument)
+#         vinfo = astate.estate[rhs]
+#     else
+#         return
+#     end
+#     add_escape_change!(astate, lhs, vinfo)
+# end
+
+# linear scan to find regions in which potential throws will be caught
+function compute_tryregions(ir::IRCode)
```

Comment:
Can this use `Core.Compiler.compute_trycatch` now?

Reply:
`Core.Compiler.compute_trycatch` somehow sometimes throws `unbalanced try/catch` assertion error when applied to `IRCode` (should be easily reproduced). It may be worth digging into what's happening there, but for escape analysis we only need to know the locations of `:enter` expressions and that analysis can be done as a simple linear scan of `compute_frameinfo`.